ide/db.sql.editor/src/org/netbeans/modules/db/sql/analyzer/SQLStatementAnalyzer.java

```diff
@@ -239,6 +239,8 @@ protected QualIdent parseIdentifier() {
     }
 
     protected String getUnquotedIdentifier() {
+        if (quoter == null)
+            return "";
```

Comment:
Silly question, how should that "Quoter" be created? 

I see a case like the below using SQLIdentifer to created it, but in this case since there is no conn available, using the meta data I don't believe would be possible. 

DatabaseMetaData dmd = conn.getMetaData();
quoter = SQLIdentifiers.createQuoter(dmd)

I also see 
NonASCIIQuoter(String quoteString)

So would quoter = new NonASCIIQuoter(""); be what you're suggesting or something else?

Reply:
It depends. I was the one, that pointed out, that it is highly problematic to try to create a completion without a corresponding connection. There is Standard SQL, so you can build a quoter based on that, but then you also need to guess what is considered a keyword an d what not.

Maybe there is a  different approach possible: Don't try to be clever and use a parser, but use keyword based completion. You could have a look at phpmyadmin completion how it is handled there. I doubt, that they went the way of a full parser in JS, but most probably complete based on keyword prefix matching.

Reply:
As a new review was requested: this is still broken.

Reply:
Just for clarification, what is meant by "Standard SQL" in this context?  Are you talking about a specific class, a "new class", or referring to "standard SQL syntax"?

Reply:
What is meant by phpmyadmin?  I can't find a file or project with that.  Is that in the php.editor or some plugin?

Reply:
Yes SQL defines default quoting behavior - from the top of my head:

- identifiers are quoted with double quotes: `"IDENTIFIER"`
- strings are quoted with single quotes: `'STRING'`
- escaping the quote itself is handled by doubling itself `""""` is the Identifier `"` and `''''` is the string `'`

Sounds simple - it is not because DB vendors screwed up. I remember, that MySQL has some more creative quotes and escapes, which make "MySQL SQL" incompatible" with "Standard SQL".

phpmyadmin is a popular PHP based tool for accessing mysql databases. It is often found in hosted setups. In recent version I saw some quick code completion, that could not have been grammar based, but might be an alternative.

Reply:
> phpmyadmin is a popular PHP based tool for accessing mysql databases. It is often found in hosted setups. In recent version I saw some quick code completion, that could not have been grammar based, but might be an alternative.

Little rusty on my php and don't have that setup at the moment to try this but from what I can tell there is [SqlAutoCompleteController file used for sql autocomplete ](https://github.com/phpmyadmin/phpmyadmin/blob/master/libraries/classes/Controllers/Database/SqlAutoCompleteController.php) which appears to basically creates an array of autocomplete items of db and table name related attributes.  I don't see the basic sql side of things which may be elsewhere so I will poke around some more.

Reply:
> Yes SQL defines default quoting behavior - from the top of my head:
> 
> * identifiers are quoted with double quotes: `"IDENTIFIER"`
> * strings are quoted with single quotes: `'STRING'`
> * escaping the quote itself is handled by doubling itself `""""` is the Identifier `"` and `''''` is the string `'`
> 
> Sounds simple - it is not because DB vendors screwed up. I remember, that MySQL has some more creative quotes and escapes, which make "MySQL SQL" incompatible" with "Standard SQL".

Looking at one of the [w3 resources SQL syntax page in identifier section](https://www.w3resource.com/sql/sql-syntax.php#IDENTIF) it does look more complicated.

Reply:
> As an alternative, in that code it uses the quoter to "unquote" stuff, but if a quoter is not available (due to lack of connection), could it just return the text as is with something like?
> 
> ```
> if (quoter == null)
> return seq.token().text().toString();
> ```

Is this more acceptable then returning the "" mentioned in the original comment?

Reply:
Yes, there is a reason the connection object is required - do you gain anything by even calling into this when no connection is available? I think I mentioned it earlier, maybe refitting code, that requires a connection to a connection less world is not the right approach.

Reply:
That said, looking at the w3 reference above, I noticed after "SELECT" there can be optionally "DISTINCT | ALL" and can be "*" or select List (i.e. identifiers).  Not sure if the handling of DISTINCT, ALL, and/or * might need to be added at some point.  You think maybe a new issue for adding that would be worth wild?

Reply:
> Looking at one of the [w3 resources SQL syntax page in identifier section](https://www.w3resource.com/sql/sql-syntax.php#IDENTIF) it does look more complicated.

The w3 reference mentions after "SELECT" there can be optionally "DISTINCT | ALL" and either a wildcard ("*") or select List (i.e. identifiers). Unless I've missed it, I'm not sure these cases are handled.  Is a new issue for adding that worth wild?

Reply:
When using another tool for SQL it allows user to write without a connection but cannot execute sql or do autocomplete without it.

Reply:
Unless I'm assuming incorrectly, when no connection is active, then assume no real need to quote/unquote any table/column would be necessary, so is going with the below reasonable? 

```
if (quoter == null)
return seq.token().text().toString();
```

Reply:
My pain point: Instead of trying to fix logic, that relies hard on a connection, why not refactor this to split connection less and connction required codepaths, then it would be clear and you'd not need to unquote.

Reply:
Matthias, I appreciate the feedback.  As a new commiter, I know I;m not there yet but will get there with help.

The original scope of this ticket was...if no connection, allow some form of auto completion without connection based details and avoid the connection dialog from preventing futher autocompletion.   As it is now I think it does this.  

The remaining concenrs is how getUnquotedIdentifier behaves when no connection is available.  Initial changes was to returned "", which I realize is wrong, risking the loss of given seq.token.txt details.  Next change was to return the results without being "unquoted" with given "return seq.token().text().toString();"; whether it needed or not was dependent on the database in use which if not available is where problems may occur.  When quoter is not available or used avoids risk of a null pointers and return the sequence of text in the given context without being concerned without being "unquoted" independent of whether a "quoter" was available or not.

When quoter is available and calls the "unquote" method found in the SQLIdentifer.Quoter abstrace class which is implemented/extended in the DatabaseMetaDataQuoter implementation; if quoter was set correctly previously (when DB is available) then instance of quoter should be available and all is right with the world.  However, when quoter is unavailable (not connection and not setup earlier) then something different is needed.  

So is the change suggetion to implement a new ConnectionLessQuoter or BasicQuoter under the SqlIdentifer class and use this when connection is not available earlier in the flow?

It seems many of the methods in the Quoter abstract class probably should be static methods.  So is the refactoring mentioned involve moving some of these out into a "static" context in some way?

As implemented, "getUnquotedIdentifer" is used in placed like "InsertStatementAnalyzer, SQLStatemenAnalyzer, and SelectStatementAnalyzer, analyzeChosenColumns, etc.  In other places, some form of interface is used like "TablesClause" (assume this would be "connection based") used in places such as "DeleteStatementAnalyzer", InsertStatementAnalyzer, SelectStatementAnalyzer, and UpdateStatementAnalyzer".  Or uses a call to "parseIdentifier" which also use the getUnquotedIdentifer or parseTableIdent to return table names "Identifiers".  So are changes to standardize these across context needed here?

For SqlStatementAnalyizer, the "connection" state is not know directly, but is available previously such as in the SQLCompletionQuery context where it established the quoter and passes forward to applicable as.  So in a given Analyzer context, the "quoter" was the way to determine if there was or wasn't a connection without passing around the connections from other scope/context.

Do this type of aspect reflect the "connection vs non-connection" based changes preferred in earlier comment?

Connection may be needed when determining what kind of identifiers to include in the autocomplete options in applicable context of a given sql statement (i.e. between select xxx from, insert .. INTO xxxx, etc.) but not others (i.e. empty line so use initial SQL keywords).  WHen available then the identifier such as table and columns identifiers based on the connection are possible which I bleive depneds on connected based DatabaseMetaDataQuoter  and related classed.  This may help determine if  should or shouldn't quote given identifiers (database/schema/table/column) for given DB product.  If connection is not available, then maybe able to add "non connection based identifier items" (i.e. "*", Count(*), DISTINCT, etc. - not sure it does this presently; is this where a StandardSQLQuoter might come in to lay?).  

In each call to "getUnquotedIdentifer()" it return a list Strings with possible items.  It iterates through each token in the current expression and may key on before/after "dot" or if typing an identifier make another call.  In other cases it may use the above mentioned "createTable" type style so

Can't really create a DB quoter without a give connection; and no easy way to determine what connection to use without user interaction (i.e. selecting DB connection if available, creating a connection, etc.).  Some sort of "default" connection (probably set in the connection pulldown) could be persisted in some way and reused after it was first established to reduce the chance it's not setup, but that also seems like a different type of change (i.e. maybe a new "store selected connection and reused next time file or SQL editor is opened" type ticket).

Reply:
Have raises separate tickets to capture refactor changes for connection vs connectionless cases.
NETBEANS-5831 Create a SQL Standard Quoter for Use with Connectionless Cases

Reply:
In the case where connectionless and no quoter is available in the context of getUnquotedIdentifier, returning seq.token().text().toString().

No unquoting is really possible unless a given connection to a specific DB is available to determine how to "unquote" things for the given DB (connection which is not available).  

Some of this behavior could changes when implementing NETBEANS-5831 to provide some form of "connection less" unquoting.